optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,134 @@ private void initBenchmarkDirectoryAndSubdirs() {
+                                SolverBenchmarkResult[] solverBenchmarkResultArray,
+                                long timeLeftPerSolverConfig) {
+        for (SolverBenchmarkResult solverBenchmarkResult : solverBenchmarkResultArray) {
             TerminationConfig originalTerminationConfig = solverBenchmarkResult.getSolverConfig().getTerminationConfig();
             TerminationConfig tmpTerminationConfig = originalTerminationConfig == null
                     ? new TerminationConfig() : originalTerminationConfig.clone();
-            tmpTerminationConfig.shortenTimeMillisSpentLimit(timeLeft);
+            tmpTerminationConfig.overwriteTimeMillisSpentLimit(timeLeftPerSolverConfig);
             solverBenchmarkResult.getSolverConfig().setTerminationConfig(tmpTerminationConfig);
 
+            Integer singleBenchmarkResultIndex = singleBenchmarkResultIndexMap.get(solverBenchmarkResult);
+            if (singleBenchmarkResultIndex == null ||
+                    singleBenchmarkResultIndex >= solverBenchmarkResult.getSingleBenchmarkResultList().size()) {
+                singleBenchmarkResultIndex = 0;
+            }
+            SingleBenchmarkResult singleBenchmarkResult
+                    = solverBenchmarkResult.getSingleBenchmarkResultList().get(singleBenchmarkResultIndex);
+            if (singleBenchmarkResult != null) {
+                SingleBenchmarkRunner singleBenchmarkRunner = new SingleBenchmarkRunner(singleBenchmarkResult);
+                Future<SingleBenchmarkRunner> future = warmUpExecutorCompletionService.submit(singleBenchmarkRunner);
+                futureMap.put(future, singleBenchmarkRunner);
+                singleBenchmarkResultIndexMap.put(solverBenchmarkResult, singleBenchmarkResultIndex + 1);
+            }
+        }
+    }
+
+    private void warmUp(Map<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner> futureMap,
+                        ConcurrentMap<SolverBenchmarkResult, Integer> singleBenchmarkResultIndexMap, long timePhaseEnd) {
+        // Wait for the warm up benchmarks to complete
+        int tasksCount = futureMap.size();
+        for (int i = 0; i < tasksCount; i++) {
+            Future<SingleBenchmarkRunner> future;
             try {
-                Solver solver = solverBenchmarkResult.getSolverConfig().buildSolver();
-                solver.solve(problemBenchmarkResult.readPlanningProblem());
-            } catch (RuntimeException e) {
-                singleBenchmarkRunnerExceptionLogger.warn("The warmUp of singleBenchmark ("
-                        + singleBenchmarkResult.getName() + ") failed.", e);
-                // TODO update firstFailureSingleBenchmarkRunner (when warmups happen in a Runner too in parallel)
+                future = warmUpExecutorCompletionService.take();
+            } catch (InterruptedException e) {
+                singleBenchmarkRunnerExceptionLogger.error("Waiting for a warm up singleBenchmarkRunner was interrupted.", e);
+                continue;
```

Comment:
Options/opinions on how to handle this correctly:

* A: Logging and continuing
```java
logger.error("Waiting for a warm up singleBenchmarkRunner was interrupted.", e);
continue;
```

* B: Throwing the InterruptedException to a higher level
  * Widely recommended, but does it make sense in our case?
  * What is it that we want to tell the caller?

* C: Setting the interrupted status and returning from the method
  * Widely recommended, but does it make sense in our case?
  * Do we actually tell the caller anything? What does returning too soon mean to it?

```java
logger.error("Waiting for a warm up singleBenchmarkRunner was interrupted.", e);
Thread.currentThread().interrupt();
return;
```

* D: Throwing a RuntimeException and failing (too?) fast

```java
throw new RuntimeException("Waiting for a warm up singleBenchmarkRunner was interrupted.", e);
```

* E: Actively waiting (doesn't throw an InterruptedException)
  * Uses the CPU extensively for an uninteresting and not needed task

```java
while (future == null) {
    future = warmUpExecutorCompletionService.poll();
}
```